In [1]:
import pandas as pd

## Reading files

In [2]:
wh2018 = pd.read_csv('2018_v2.csv')
wh2019 = pd.read_csv('2019_v2.csv')

Adding extra column to identify year of population

In [3]:
wh2018['year'] = 2018
wh2019['year'] = 2019

Union of three files

In [4]:
wh = pd.concat([wh2018,wh2019])

In [5]:
wh.columns

Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption', 'year'],
      dtype='object')

## Ejercicios

1. Partiendo de Wh, realice un analisis de valores nulos. que porcentaje de valores nulos se encuentran en cada columna por anio. presente el analisis en un dataframe

Ejemplo:

In [6]:
pd.DataFrame({
    "year": [2017,2017]
    , "column" : ["Happiness.Rank","Whisker.low"]
    , "percentage.of.nans" : [0.2,0.2]
})

year          column  percentage.of.nans
0  2017  Happiness.Rank                 0.2
1  2017     Whisker.low                 0.2

### resolucion ejercicio uno

In [39]:
wh.isnull().sum().sum()/wh.shape[0]

0.4166666666666667

2. partiendo de Wh, por año muestre un resumen de todas las columnas.

nota: aunque solo tenemos tres años, que pasaria si en nuestro dataframe tenemos 20 años? resuelva su ejercicio lo mas dinamico que se pueda

### resolucion ejercicio dos

In [74]:
wh1= wh.iloc[:,:].groupby(['year']).count()

In [75]:
wh1

Overall rank  Country or region  Score  GDP per capita  Social support  \
year                                                                           
2018           156                156    151             147             141   
2019           156                156    142             153             155   

      Healthy life expectancy  Freedom to make life choices  Generosity  \
year                                                                      
2018                      144                           144         139   
2019                      142                           145         152   

      Perceptions of corruption  
year                             
2018                        149  
2019                        150

3. partiendo de Wh, remplace los valores NaN de las columnas por su mean, el mean debe de ser calculado por año para ser mas exacto

### resolucion ejercicio tres

In [77]:
wh.isnull().sum()

Overall rank                     0
Country or region                0
Score                           19
GDP per capita                  12
Social support                  16
Healthy life expectancy         26
Freedom to make life choices    23
Generosity                      21
Perceptions of corruption       13
year                             0
dtype: int64

In [80]:
z=wh.iloc[:,:].groupby(['year']).mean()

Overall rank     Score  GDP per capita  Social support  \
year                                                           
2018          78.5  5.354060        0.892748        1.205915   
2019          78.5  5.434028        0.900758        1.208265   

      Healthy life expectancy  Freedom to make life choices  Generosity  \
year                                                                      
2018                 0.597576                      0.456514    0.181360   
2019                 0.722965                      0.395441    0.185092   

      Perceptions of corruption  
year                             
2018                   0.112859  
2019                   0.109053

In [88]:
wh['Score'] = wh['Score'].fillna(wh['Score'].mean())

4. Partiendo de Wh limpio, para cada metrica por pais, muestra la diferencia del ultimo valor encontrado contra el primero. 

Ejemplo:

In [62]:
wh.loc[ wh['Country or region'] == 'Switzerland' , : ]

Overall rank Country or region  Score  GDP per capita  Social support  \
4             5       Switzerland  7.487           1.420           1.549   
5             6       Switzerland  7.480           1.452           1.526   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
4                    0.927                         0.660       0.256   
5                    1.052                         0.572       0.263   

   Perceptions of corruption  year  
4                      0.357  2018  
5                      0.343  2019

Como se puede observar para Switzerland en Score la diferencia seria 7.480 - 7.487

Claramente en su solucion la columna año se debera de eliminar

### resolucion ejercicio cuatro

In [103]:
z=wh.groupby(['Country or region'], as_index=False).agg( {'Score':['max','min']})

In [104]:
z

Country or region     Score       
                            max    min
0         Afghanistan  3.632000  3.203
1             Albania  4.719000  4.586
2             Algeria  5.295000  5.211
3              Angola  3.795000  3.795
4           Argentina  6.388000  6.086
..                ...       ...    ...
155         Venezuela  4.806000  4.707
156           Vietnam  5.175000  5.103
157             Yemen  5.392816  3.355
158            Zambia  5.392816  4.377
159          Zimbabwe  3.692000  3.663

[160 rows x 3 columns]

In [105]:
z.columns

MultiIndex([('Country or region',    ''),
            (            'Score', 'max'),
            (            'Score', 'min')],
           )

In [110]:
z=wh.groupby(['Country or region','year'], as_index=False).max()

In [113]:
x=wh.groupby(['Country or region','year'], as_index=False).min()

In [114]:
x.loc[]

Country or region  year  Overall rank     Score  GDP per capita  \
0         Afghanistan  2018           145  3.632000           0.332   
1         Afghanistan  2019           154  3.203000           0.350   
2             Albania  2018           112  4.586000           0.916   
3             Albania  2019           107  4.719000           0.947   
4             Algeria  2018            84  5.295000           0.979   
..                ...   ...           ...       ...             ...   
307             Yemen  2019           151  5.392816           0.287   
308            Zambia  2018           125  4.377000           0.562   
309            Zambia  2019           138  5.392816           0.578   
310          Zimbabwe  2018           144  3.692000           0.357   
311          Zimbabwe  2019           146  3.663000           0.366   

     Social support  Healthy life expectancy  Freedom to make life choices  \
0             0.537                    0.255                         0.085   
1             0.517                    0.361                         0.000   
2             0.817                    0.790                         0.419   
3             0.848                    0.874                         0.383   
4             1.154                    0.687                         0.077   
..              ...                      ...                           ...   
307           1.163                    0.463                         0.143   
308           1.047                    0.295                         0.503   
309           1.058                    0.426                         0.431   
310           1.094                    0.248                         0.406   
311           1.114                      NaN                         0.361   

     Generosity  Perceptions of corruption  
0         0.191                      0.036  
1         0.158                      0.025  
2         0.149                      0.032  
3         0.178                      0.027  
4         0.055                      0.135  
..          ...                        ...  
307       0.108                      0.077  
308       0.221                      0.082  
309       0.247                        NaN  
310       0.132                      0.099  
311       0.151                      0.089  

[312 rows x 10 columns]